README
NAME
    Apache2::WebApp::Extra::Admin - Web based admin control panel

SYNOPSIS
    This module should NOT be used directly.

DESCRIPTION
    This package provides application sources for a fully functional web
    based admin Control Panel. You can easily add this *Extra* to an
    existing project using the "webapp-extra" helper script provided with
    the WebApp::Toolkit

    http://cpansearch.perl.org/src/MBROOKS/Apache2-WebApp-Extra-Admin-0.15/preview

  Features
    * Extensible XHTML/CSS interface.
    * Template based output.
    * Compatible with Internet Explorer, Firefox, Safari, Chrome, and Opera
    web browsers.
    * User management panel.
    * User action logging.
    * Result sorting capabilities.
    * Pagination (Next/Last pages)

PREREQUISITES
      Apache2::WebApp
      Apache2::WebApp::Plugin::DateTime
      Apache2::WebApp::Plugin::DBI (optional)
      Apache::Htpasswd
      Params::Validate

INSTALLATION
    From source:

      $ tar xfz Apache2-WebApp-Extra-Admin-0.X.X.tar.gz
      $ perl MakeFile.PL PREFIX=~/path/to/custom/dir LIB=~/path/to/custom/lib
      $ make
      $ make install

    Perl one liner using CPAN.pm:

      $ perl -MCPAN -e 'install Apache2::WebApp::Extra::Admin'

    Use of CPAN.pm in interactive mode:

      $ perl -MCPAN -e shell
      cpan> install Apache2::WebApp::Extra::Admin
      cpan> quit

    Just like the manual installation of Perl modules, the user may need
    root access during this process to insure write permission is allowed
    within the installation directory.

GETTING STARTED
  HELPER SCRIPT
   Export project settings to the Unix shell
      $ source /path/to/project/.projrc

   Add this *Extra* to an existing project
      $ webapp-extra Admin

   Restart your application
      $ webapp-kickstart

   Accessing the admin control panel
      http://domain.com/app/admin

CONFIGURATION
    Unless it already exists, add the following to your projects
    *webapp.conf*

      [custom]
      items_per_page = 25     # default value

SECURITY
    By default, the admin control panel is public and not secure. You can
    enable user access controls by adding the following *Apache* directives
    into your project "httpd.conf"

      <Location /app/admin>
          AuthType Basic
          AuthName "Admin"
          AuthUserFile /home/webapp/conf/htpasswd
          Require valid-user
      </Location>

    As always, you must restart your application for changes to take effect.

ADMIN ACCESS
    The login information below is currently set-up by default. In order to
    manage user access this account must exist within your "htpasswd"

      User Name       admin
      Password        password

OPTIONAL
    If database support is available, you can log control panel user actions
    to a database. The database table to store logging information is
    auto-generated. Since this is the case, the SQL user must have CREATE
    privileges.

    In case you want to manually create this table, you can run the
    following SQL statement on the database command-line:

      CREATE TABLE IF NOT EXISTS admin_log (
          admin_log_id int(3) NOT NULL AUTO_INCREMENT,
          admin_log_username varchar(15) DEFAULT NULL,
          admin_log_action varchar(55) DEFAULT NULL,
          admin_log_query_string text,
          admin_log_user_ip_addr varchar(15) DEFAULT NULL,
          admin_log_created datetime DEFAULT NULL,
          PRIMARY KEY (admin_log_id)
        );

    If a database is not available at the time data is collected, this
    information will be written to a file that can be located in your
    project "logs" directory.

DEBUG MODE
    If debugging is enabled, the URI and query string are logged. This can
    pose as a security risk when running in a production environment since
    personal information (including passwords) may be exposed.

SEE ALSO
    Apache2::WebApp, Apache2::WebApp::Plugin::DBI,
    Apache2::WebApp::Plugin::DateTime, Apache::Htpasswd

AUTHOR
    Marc S. Brooks, <mbrooks@cpan.org> <http://mbrooks.info>

COPYRIGHT
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://dev.perl.org/licenses/artistic.html>

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.



Changes
Revision history for Perl module Apache2::WebApp::Extra::Admin

0.16
    - Migrated project SCM and code repository to Google Project hosting.
    - Updated POD and README artistic license URL in COPYRIGHT clause.
    - Changed each module version so that I can verify that the PAUSE packager/uploader script works as expected.
    - Updated license field in META.yml to fix 'License Unknown' issue on CPAN
    - Updated PREREQ_PM module versions in Makefile.PL
    - Updated module versions in META.yml requires field.

0.17
    - Added $(PREFIX) in Makefile.PL to build script template install path.

Makefile.PL
use ExtUtils::MakeMaker;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile(
    NAME         => 'Apache2::WebApp::Extra::Admin',
    VERSION_FROM => 'lib/Apache2/WebApp/Extra/Admin.pm', # finds \$VERSION
    AUTHOR       => 'Marc S. Brooks (mbrooks@cpan.org)',
    PREREQ_PM    => {
       'Apache2::WebApp'                   => 0.38,
       'Apache2::WebApp::Plugin::DateTime' => 0.06,
       'Apache2::WebApp::Plugin::DBI'      => 0.09,
       'Apache::Htpasswd'                  => 1.8,
       'Params::Validate'                  => 0,
    }
  );

# Install build script template dependencies (non-standard way)
sub MY::postamble {
'
install::
	cp -rv usr/share/webapp-toolkit/extra* $(PREFIX)/share/webapp-toolkit
';
}

usr/share/webapp-toolkit/extra/htdocs/admin/js/form.js
/*
 *  Apache2::WebApp::Toolkit - Admin Control Panel (form.js)
 *  Copyright (C) 2010 Marc S. Brooks <mbrooks@cpan.org>
 *
 *  Licensed under the terms of the BSD License
 *  http://www.opensource.org/licenses/bsd-license.php
 */

var args = [];
var form;

function requireFields() {
	form = countForms();	// form position

	if (arguments.length) {
		args = arguments;
		document.onkeyup = function (e) { allowSubmit() }
		document.onclick = function (e) { allowSubmit() }
		allowSubmit();
	}

	eventFieldError();
}

function allowSubmit() {
	if (fieldCheck()) {
		document.getElementById('allow').disabled = false;
	}
	else {
		document.getElementById('allow').disabled = true;
	}
}

function fieldCheck() {
	for ( var i = 0; i < args.length; i++ ) {
		var elm = document.forms[form].elements[args[i]];

		if (!elm.value) { return false }

		if (elm.type == 'checkbox' || elm.type == 'radio') {
			if (!elm.checked) {
				return false;
			}
		}
	}
	return true;
}

function countForms() {
	var total = 0;
	for ( var i = 0; i < document.forms.length; i++ ) {
		total++;
	}

	if (total > 0) {
		return total - 1;	// array always begins with a 0
	}
	else {
		return 0;
	}
}

function eventFieldError() {
	var elm = document.getElementsByTagName('*');
	var obj = document.getElementById('alert');

	var error;

	for ( var i = 0; i < elm.length; i++ ) {
		if (elm[i].className.match(/error/)) {
			elm[i].onmouseover = function (e) {
				error = obj.innerHTML;
				obj.innerHTML = this.title;
				this.style.backgroundColor = "#FF0000";
				this.style.color           = "#FFFFFF";
			}

			elm[i].onmouseout = function (e) {
				obj.innerHTML = error;
				this.style.backgroundColor = "#FFFFFF";
				this.style.color           = "#FF0000";
			}
		}
	}
}

function eventSubmit() {
	document.getElementById('allow').onclick = function (e) {
		disableSubmit()
	};
}

function disableSubmit() {
	document.getElementById('allow').disabled = true;
	document.forms[form].submit();	// i.e. is lame, fix
}

function focusFirstField(name) {
	if (!name) { name = 0 }

	for( var i = 0; i < document.forms[name].length; i++ ) {
		if (document.forms[name][i].type == 'text' || document.forms[name][i].type == 'password' ) {
			if (document.forms[name][i].disabled != true) {
				document.forms[name][i].focus();
				break;
			}
		}
	}
}


usr/share/webapp-toolkit/extra/htdocs/admin/js/menu.js
/*
 *  Apache2::WebApp::Toolkit - Admin Control Panel (menu.js)
 *  Copyright (C) 2010 Marc S. Brooks <mbrooks@cpan.org>
 *
 *  Licensed under the terms of the BSD License
 *  http://www.opensource.org/licenses/bsd-license.php
 */

var menu_total;

function initMenus(total) {
	var last = getCookie('last');

	menu_total = total;

	if (last) {
		viewOptions(last);
	}
	else {
		if (total < 2) {    // Control Panel & About menus
			resetMenu();
		}
		else {
			viewOptions('menu_1');
		}
	}
}

function loadMenu(mode) {
	var last = getCookie('last');
    
	if (last != mode) {
        parent.side.viewOptions(mode);
    }
}

function resetMenu() {
	for ( var i = 1; i <= menu_total; i++ ) {
		document.getElementById( 'menu_' + i ).style.display = 'none';
	}
}

function viewOptions(mode) {
	var obj = document.getElementById(mode);

	if (obj.style.display == 'block') {
		obj.style.display = 'none';

		delCookie('last');
	}
	else {
		resetMenu();

		obj.style.display = 'block';

		setCookie('last', mode);
	}
}


usr/share/webapp-toolkit/extra/htdocs/admin/js/cookie.js
/*
 *  Apache2::WebApp::Toolkit - Admin Control Panel (cookie.js)
 *  Copyright (C) 2010 Marc S. Brooks <mbrooks@cpan.org>
 *
 *  Licensed under the terms of the BSD License
 *  http://www.opensource.org/licenses/bsd-license.php
 */

function getCookie(name) {
	var obj = document.cookie;
	var arg = name + "=";
	var beg = obj.indexOf("; " + arg);

	if (beg == -1) {
		beg = obj.indexOf(arg);

		if (beg != 0) { return null };
	}
	else {
		beg += 2;
	}

	var end = document.cookie.indexOf(";", beg);

	if (end == -1) {
		end = obj.length;
	}

	return unescape(obj.substring(beg + arg.length, end));
}

function setCookie( name, value, expires, path, domain, secure ) {
	document.cookie = name + "="  + escape(value) +
		((expires) ? "; expires=" + expires.toGMTString() : "") +
		((path)    ? "; path="    + path                  : "") +
		((domain)  ? "; domain="  + domain                : "") +
		((secure)  ? "; secure"                           : "");
}

function delCookie(name) {
	var val = getCookie(name);
	var exp = new Date();

	exp.setTime (exp.getTime() - 1);

	document.cookie = name + "=" + val + "; expires=" + exp.toGMTString();
}


lib/Apache2/WebApp/Extra/Admin.pm
#----------------------------------------------------------------------------+
#
#  Apache2::WebApp::Extra::Admin - Web based admin control panel
#
#  DESCRIPTION
#  This package provides application sources for a fully functional web based
#  admin Control Panel.  You can easily add this Extra to an existing project
#  using the webapp-extra helper script provided with the WebApp::Toolkit
#
#  AUTHOR
#  Marc S. Brooks <mbrooks@cpan.org>
#
#  This module is free software; you can redistribute it and/or
#  modify it under the same terms as Perl itself.
#
#----------------------------------------------------------------------------+

package Apache2::WebApp::Extra::Admin;

our $VERSION = 0.18;

1;

__END__

=head1 NAME

Apache2::WebApp::Extra::Admin - Web based admin control panel

=head1 SYNOPSIS

This module should NOT be used directly.

=head1 DESCRIPTION

This package provides application sources for a fully functional web based
admin Control Panel.  You can easily add this I<Extra> to an existing project
using the C<webapp-extra> helper script provided with the WebApp::Toolkit

L<http://cpansearch.perl.org/src/MBROOKS/Apache2-WebApp-Extra-Admin-0.15/preview>

=head2 Features

=over

=item * Extensible XHTML/CSS interface.

=item * Template based output.

=item * Compatible with Internet Explorer, Firefox, Safari, Chrome, and Opera web browsers.

=item * User management panel.

=item * User action logging.

=item * Result sorting capabilities.

=item * Pagination (Next/Last pages)

=back

=head1 PREREQUISITES

  Apache2::WebApp
  Apache2::WebApp::Plugin::DateTime
  Apache2::WebApp::Plugin::DBI (optional)
  Apache::Htpasswd
  Params::Validate

=head1 INSTALLATION

From source:

  $ tar xfz Apache2-WebApp-Extra-Admin-0.X.X.tar.gz
  $ perl MakeFile.PL PREFIX=~/path/to/custom/dir LIB=~/path/to/custom/lib
  $ make
  $ make install

Perl one liner using CPAN.pm:

  $ perl -MCPAN -e 'install Apache2::WebApp::Extra::Admin'

Use of CPAN.pm in interactive mode:

  $ perl -MCPAN -e shell
  cpan> install Apache2::WebApp::Extra::Admin
  cpan> quit

Just like the manual installation of Perl modules, the user may need root access during
this process to insure write permission is allowed within the installation directory.

=head1 GETTING STARTED

=head2 HELPER SCRIPT

=head3 Export project settings to the Unix shell

  $ source /path/to/project/.projrc

=head3 Add this I<Extra> to an existing project

  $ webapp-extra Admin

=head3 Restart your application

  $ webapp-kickstart

=head3 Accessing the admin control panel

  http://domain.com/app/admin

=head1 CONFIGURATION

Unless it already exists, add the following to your projects I<webapp.conf>

  [custom]
  items_per_page = 25     # default value

=head1 SECURITY

By default, the admin control panel is public and not secure.  You can 
enable user access controls by adding the following I<Apache> directives
into your project C<httpd.conf>

  <Location /app/admin>
      AuthType Basic
      AuthName "Admin"
      AuthUserFile /home/webapp/conf/htpasswd
      Require valid-user
  </Location>

As always, you must restart your application for changes to take effect.

=head1 ADMIN ACCESS

The login information below is currently set-up by default.  In order to
manage user access this account must exist within your C<htpasswd>

  User Name       admin
  Password        password

=head1 OPTIONAL

If database support is available, you can log control panel user actions
to a database.  The database table to store logging information is 
auto-generated.  Since this is the case, the SQL user must have CREATE 
privileges.

In case you want to manually create this table, you can run the following
SQL statement on the database command-line:

  CREATE TABLE IF NOT EXISTS admin_log (
      admin_log_id int(3) NOT NULL AUTO_INCREMENT,
      admin_log_username varchar(15) DEFAULT NULL,
      admin_log_action varchar(55) DEFAULT NULL,
      admin_log_query_string text,
      admin_log_user_ip_addr varchar(15) DEFAULT NULL,
      admin_log_created datetime DEFAULT NULL,
      PRIMARY KEY (admin_log_id)
    );

If a database is not available at the time data is collected, this 
information will be written to a file that can be located in your project 
C<logs> directory.

=head1 DEBUG MODE

If debugging is enabled, the URI and query string are logged.  This can
pose as a security risk when running in a production environment since
personal information (including passwords) may be exposed.

=head1 SEE ALSO

L<Apache2::WebApp>, L<Apache2::WebApp::Plugin::DBI>,
L<Apache2::WebApp::Plugin::DateTime>, L<Apache::Htpasswd>

=head1 AUTHOR

Marc S. Brooks, E<lt>mbrooks@cpan.orgE<gt> L<http://mbrooks.info>

=head1 COPYRIGHT

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://dev.perl.org/licenses/artistic.html>

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.

=cut
